### DISCUSSION TITLE
Highlight diff between hcl / terraform

### DISCUSSION THREAD

**Discussion by rodhash** [nvim-treesitter/nvim-treesitter#4265] [General]
Hi,

Recently noticed some differences between hcl / terraform parsers .. while hcl is working just fine, terraform is kinda weird

ps: Other lang/parsers work just fine such as python, cpp, etc.

Running latest versions:
Nvim `v0.9.0-dev-865`
Nvim treesitter 660a070a
Mac M1 (brew updated)

`checkhealth nvim-treesitter`

```v
==============================================================================
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- OK `tree-sitter` found 0.20.7 (439a4804db27988e2e3b5851f9d19ad495aabe2e) (parser generator, only needed for :TSInstallFromGrammar)
- OK `node` found v19.5.0 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: Apple clang version 14.0.0 (clang-1400.0.29.202)
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "arm64",
  release = "22.2.0",
  sysname = "Darwin",
  version = "Darwin Kernel Version 22.2.0: Fri Nov 11 02:04:44 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T8103"
} ~

Parser/Features         H L F I J
  - cpp                 ✓ ✓ ✓ ✓ ✓
  - bash                ✓ ✓ ✓ . ✓
  - html                ✓ ✓ ✓ ✓ ✓
  - hcl                 ✓ . ✓ ✓ ✓
  - v                   ✓ ✓ ✓ ✓ ✓
  - yaml                ✓ ✓ ✓ ✓ ✓
  - typescript          ✓ ✓ ✓ ✓ ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - javascript          ✓ ✓ ✓ ✓ ✓
  - lua                 ✓ ✓ ✓ ✓ ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~

ps: terraform parser was installed after saving this output
```

**Comment by rodhash** [comment]
actually hcl is working fine, the problem was terraform that was missing from my auto install setup line

**Comment by rodhash** [comment]
actually hcl parser is working way better than terraform one..

terraform is the one not highlighting some keywords .. if I rename the file from .tf to .hcl it all works fine ..

I'm wondering if I can force the hcl to be used instead of terraform

**Comment by rodhash** [comment]
Tried changing the filetype to terraform.hcl .. it worked, the highlight in place was hcl which is what I wanted, good

but even tho terraform server attached to the buffer, it's not providing completion and other stuff

writting a basic config:
```v
local server_opts = {}
server_opts.filetypes = { "terraform.hcl", "terraform", "hcl", "tf", }
require("lspconfig").terraformls.setup(server_opts)
```

curious thing is, if I edit some .tf file then server attaches and also works

**Comment by theHamsta** [comment]
Install the `terraform` parser. It was forked from the `hcl` parser.

**Comment by rodhash** [comment]
Yes actually I installed, it was after I run the initial checkhealth, now I do have it installed:

```v
Parser/Features         H L F I J
  - lua                 ✓ ✓ ✓ ✓ ✓
  - html                ✓ ✓ ✓ ✓ ✓
  - terraform           ✓ . ✓ ✓ ✓
  - hcl                 ✓ . ✓ ✓ ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - bash                ✓ ✓ ✓ . ✓
  - yaml                ✓ ✓ ✓ ✓ ✓
  - javascript          ✓ ✓ ✓ ✓ ✓
  - typescript          ✓ ✓ ✓ ✓ ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - v                   ✓ ✓ ✓ ✓ ✓
  - cpp                 ✓ ✓ ✓ ✓ ✓
```

but for some reason, this one is not highlighting some keywords

**Comment by rodhash** [comment]
Here a sample recording .. When editing hcl file it works just fine, but when I move to the 2nd window terraform highlight doesn't seem to work as hcl does.

Below I have one tmux window opened for each file to show the differences between them... 1st window hcl and 2nd one terraform

apparently terraform parser doesn't apply couple hl groups such as String and Field.


ps: This is a theme that I'm working on .. but the same behaviour I noticed on others such as gruvbox:


https://user-images.githubusercontent.com/29671981/216817427-b9cb7cc1-3b09-4e69-ba3d-a15cfbbf9d79.mp4

**Reply by amaanq** [reply]
Are terraform and hcl effectively the same?

**Reply by rodhash** [reply]
sorry delay .. I'm not quite sure if they are effectivly the same.. but I do see common highlight groups on both.

From my perspective it seems that it doesn't apply colors as expected.. in this example, TreeSitter points to the group "@string", but it's not applied:
<img width="801" alt="image" src="https://user-images.githubusercontent.com/29671981/226389060-7d46fd96-c05b-4301-8a23-2ab1175215d1.png">

In here we can see the color of "@string" which is yellow:
<img width="551" alt="image" src="https://user-images.githubusercontent.com/29671981/226389468-9d264a07-1d17-4e0b-971f-72f2eb5c15fc.png">

**Comment by rodhash** [comment]
Let me give one more example .. now I'm using colorscheme Gruvbox:

TS says "@string":
<img width="597" alt="image" src="https://user-images.githubusercontent.com/29671981/226397527-0cb94dc7-dfe5-402f-852a-3c85446bbdf2.png">


But we can see that in here, "@string" is set to yellow-ish (or perhaps green-ish, my eyes are tired right now :)
<img width="532" alt="image" src="https://user-images.githubusercontent.com/29671981/226397634-8b52943c-470c-4055-9e4d-a8fe918cb4f4.png">

**Comment by rodhash** [comment]
I just figured that, even tho TS shows "@string" the actual highlight group being applied there is "Constant":


https://user-images.githubusercontent.com/29671981/226398547-9521eb93-d312-43df-8a86-3d6c67c0b2bc.mov
